crowbar_framework/app/models/service_object.rb

```diff
@@ -1408,6 +1431,35 @@ def apply_role(role, inst, in_queue)
     # XXX: This should not be done this way.  Something else should request this.
     system("sudo", "-i", Rails.root.join("..", "bin", "single_chef_client.sh").expand_path.to_s) if !ran_admin
 
+    # are there any roles to remove from the runlist?
+    # The @bcname proposal's elements key will contain the removal intentions
+    # proposal.elements =>
+    # {
+    #   "role1_remove" => ["node1"],
+    #   "role2_remove" => ["node2", "node3"]
+    # }
+    roles_to_remove = databag["deployment"][@bc_name]["elements"].select do |r|
+      r =~ /_remove/
+    end.keys
+    # returns ["role1_remove", "role2_remove"] || {}
+    roles_to_remove.each do |role_to_remove|
+      nodes_with_role_to_remove = databag["deployment"][@bc_name]["elements"][role_to_remove]
+      # returns ["node1", "node2"] || nil
+      # are there any leftover nodes where runlist removal
+      # has not been performed?
+      unless nodes_with_role_to_remove.nil?
+        # concatenate the leftover_nodes with
+        # the actual nodes that have a _remove role
+        unless leftover_nodes[role_to_remove].nil?
+          nodes_with_role_to_remove.concat(leftover_nodes[role_to_remove])
+        end
```

Comment:
@vuntz I am a bit confused. I thought the `leftover_nodes` are from a former `apply_role` which hasn't done the removal yet for some reason. So if we have an `apply_role` run without `leftover_nodes` but with a "new" removal, it won't happen, because the nodes added [here](https://github.com/crowbar/barclamp-crowbar/pull/1224/files#diff-0be8ea1f05c9441a59bb2b7ac65b1c62R1214) are not in the `leftover_nodes` hash.